test-kotlin/src/com/intellij/rt/coverage/runner.kt

```diff
@@ -29,13 +29,62 @@ fun runWithCoverage(coverageDataFile: File, testName: String, sampling: Boolean,
     return CoverageStatusTest.runCoverage(classPath, coverageDataFile, "kotlinTestData.*", "kotlinTestData.$testName.Test", sampling, calcUnloaded)
 }
 
-internal fun assertEqualsLines(project: ProjectData, expectedLines: Map<Int, String>, classNames: List<String>) {
+internal val allNames = listOf("ALL CLASSES")
+
+private fun coverageLines(project: ProjectData, classNames: List<String>): Map<Int, String> {
     val allData = ClassData("")
-    classNames
-            .map { project.getClassData(it) }
-            .forEach { allData.merge(it) }
+    if (classNames === allNames) {
+        project.classes.values.forEach { allData.merge(it) }
+    } else {
+        classNames
+                .map { project.getClassData(it) }
+                .forEach { allData.merge(it) }
+    }
     val lines = allData.getLinesData().associateBy({ it.lineNumber }, { it.status.toByte() })
-    assertEquals(expectedLines, statusToString(lines))
+    return statusToString(lines)
+}
+
+internal fun assertEqualsLines(project: ProjectData, expectedLines: Map<Int, String>, classNames: List<String>) {
+    val expected = expectedLines.toList()
+    val actual = coverageLines(project, classNames).toList()
+    var fail = false
+    compareCoverage(expected, actual, wrongLineCoverage = { i, j ->
+        System.err.println("Line ${expected[i].first}: expected ${expected[i].second} but ${actual[j].second} found")
+        fail = true
+    }, missedLine = { i ->
+        System.err.println("Line ${expected[i].first} expected with coverage ${expected[i].second}")
+        fail = true
+    }, unexpectedLine = { i ->
+        System.err.println("Unexpected line ${actual[i].first} with coverage ${actual[i].second}")
+        fail = true
+    })
+    if (fail) {
+        Assert.fail("Coverage mismatch")
```

Comment:
with this improvements you won't get IDEA's diff but would get "expected but was" lines. It may be more convenient for TC but locally, I think, diff is more useful, no? Or what kind of issue did you fix?

Reply:
This fix changes a report like this
```

expected:<{21=FULL, 23=FULL, 24=FULL, 26=FULL, 32=FULL, 33=FULL, 35=FULL, 40=FULL, 41=FULL, 42=PARTIAL, 44=FULL, 48=FULL, 49=NONE, 50=FULL}> but was:<{21=FULL, 23=FULL, 24=FULL, 26=FULL, 32=FULL, 33=FULL, 35=FULL, 40=FULL, 41=FULL, 42=PARTIAL, 44=FULL, 48=FULL, 49=FULL, 50=FULL}>
Expected :{21=FULL, 23=FULL, 24=FULL, 26=FULL, 32=FULL, 33=FULL, 35=FULL, 40=FULL, 41=FULL, 42=PARTIAL, 44=FULL, 48=FULL, 49=NONE, 50=FULL}
Actual   :{21=FULL, 23=FULL, 24=FULL, 26=FULL, 32=FULL, 33=FULL, 35=FULL, 40=FULL, 41=FULL, 42=PARTIAL, 44=FULL, 48=FULL, 49=FULL, 50=FULL}

```
to a report like this
```
Line 49: expected NONE but FULL found

Coverage mismatch
```
It is far more convenient for long files.
Diff shows exactly the same information but it needs a window to be open. So you can not see test file when looking at the diff.

Reply:
new format is convenient if there is one "bad" line, but if you would have 20 lines like `Line 49: expected NONE but FULL found` it would be not nice. 

I'd suggest to restore diff and have your output as well.

Reply:
Ok, I did it